### PR TITLE
Add emitter callback function (close #7)

### LINKED ISSUE #7: Add emitter callback function
This allows us to have a live report of how the emitter daemon is working with counts of successes and failures - useful for debugging and testing.

### PR DESCRIPTION
**Pull request by matus-tomlein**
Addresses issue #7 and adds a `set_request_callback` function to Emitter that adds a callback to be called after emit attempts.

While working on issue #54, I had to introduce some form of feedback in case events are dropped and not retried. I found out that there is no standard approach for logging in C++ and didn't want to introduce a new logging library. The best approach seemed to add a callback to Emitter that would be called in case events are dropped and let users do what they want with it. There was already an old issue for this, so I followed up on it.

I worked on this while working on issue #54 and then separated the code into 2 PRs so some things (like the changes in `HttpRequestResult`) are more related to the other issue than to adding the callback.

## Documentation: Emitter callback

The Emitter enables you to set a callback function to be called after events are attempted to be sent to the Collector. This callback is fired after HTTP requests are made and you can subscribe for specific emit statuses. The following statuses can be subscribed to:

* `EmitStatus::SUCCESS` – events were successfuly sent to the Collector.
* `EmitStatus::FAILED_WILL_RETRY` – events failed to be sent to the Collector but will be retried later.
* `EmitStatus::FAILED_WONT_RETRY` – events failed to be sent to the Collector and won't be retried later (they will be dropped).

The callback is given two arguments – list of event IDs, and their emit status. You can only set one callback at once but you can subscribe to multiple emit statuses using binary operations. The following example shows how to set a callback that is called for all three emit statuses and prints to standard output.

```cpp
emitter.set_request_callback(
    [](list<string> event_ids, EmitStatus emit_status) {
      switch (emit_status) {
      case EmitStatus::SUCCESS:
        printf("Successfuly sent %lu events.\n", event_ids.size());
        break;
      case EmitStatus::FAILED_WILL_RETRY:
        printf("Failed to send %lu events, but will retry.\n", event_ids.size());
        break;
      case EmitStatus::FAILED_WONT_RETRY:
        printf("Failed to send %lu events and won't retry.\n", event_ids.size());
        break;
      }
    },
    EmitStatus::SUCCESS | EmitStatus::FAILED_WILL_RETRY | EmitStatus::FAILED_WONT_RETRY);
```

The callback is executed in a new thread. The `set_request_callback` function can't be called when the Emitter is running.

----

There is also a [PR on data-value-resources](https://github.com/snowplow-incubator/data-value-resources/pull/84) for the documentation change.

### CODE REVIEW

**Comment by snowplowcla** [comment]
Thanks for your pull request. Is this your first contribution to a Snowplow open source project? Before we can look at your pull request, you'll need to sign a Contributor License Agreement (CLA).

:memo: Please visit https://docs.snowplowanalytics.com/docs/contributing/contributor-license-agreement/ to learn more and sign.

Once you've signed, please reply here (e.g. I signed it!) and we'll verify. Thanks.

File: src/emitter.cpp [hunk 1]
```diff
@@ -130,108 +131,161 @@ void Emitter::flush() {
...
+    for (auto const &row : event_rows) {
+      Payload event_payload = row.event;
+      event_payload.add(SNOWPLOW_SENT_TIMESTAMP, to_string(Utils::get_unix_epoch_ms()));
       string query_string = Utils::map_to_query_string(event_payload.get());
-      list<int> row_id = {it->id};
+      list<int> row_id = {row.id};
 
-      request_futures.push_back(std::async(&HttpClient::http_get, this->m_http_client.get(), this->m_url, query_string, row_id, (query_string.size() > this->m_byte_limit_get)));
-      request_futures.push_back(std::async(&HttpClient::http_get, this->m_http_client.get(), this->m_url, query_string, row_id, (query_string.size() > this->m_byte_limit_get)));
+      request_futures.push_back(async(&HttpClient::http_get, this->m_http_client.get(), this->m_url, query_string, row_id, (query_string.size() > this->m_byte_limit_get)));
```
Comment (AlexBenny) [inline src/emitter.cpp:187]: 🤔 there were two identical calls?!?
Weird we didn't notice it

File: test/emitter_test.cpp [hunk 2]
```diff
@@ -109,16 +124,16 @@ TEST_CASE("emitter") {
...
+    emitter.start();
 
-    Payload p;
-    p.add("e", "pv");
+    Payload payload;
+    payload.add("e", "pv");
 
     for (int i = 0; i < 10; i++) {
-      e.add(p);
+      emitter.add(payload);
```
Comment (AlexBenny) [inline test/emitter_test.cpp:134]: Thanks!

**Review by AlexBenny — APPROVED** [review]
LGTM!! 👍

**Pull request merged**